ott/geometry/costs.py

```diff
@@ -90,17 +90,78 @@ def tree_unflatten(cls, aux_data, children):
     return cls(*children)
 
 
+@jax.tree_util.register_pytree_node_class
+class RBFCost(CostFn):
+  """A radial-basis function cost class for translation invariant costs.
+
+  Such costs are defined using a function :math:`h`, mapping vectors to
+  real-values, to be used as:
+
+  :math:`c(x,y) = h(z)`, where :math:`z := x-y`.
+
+  If that cost function is used to form an Entropic map using the
+  :cite:`brenier:91` theorem, then the user should ensure :math:`h` is
+  strictly convex, as well as provide the Legendre transform of :math:`h`,
+  whose gradient is necessarily the inverse of the gradient of :math:`h`.
+  """
+
+  @abc.abstractmethod
+  def h(self, z: jnp.ndarray) -> float:
+    """RBF function acting on difference of `x-y` to ouput cost."""
+
+  def h_legendre(self, z: jnp.ndarray) -> float:
+    """Legendre transform of RBF function `h` (when latter is convex)."""
+    raise NotImplementedError("`h_legendre` not implemented.")
+
+  def pairwise(self, x: jnp.ndarray, y: jnp.ndarray) -> float:
+    """Evaluate h on difference between x and y."""
+    return self.h(x - y)
+
+
+@jax.tree_util.register_pytree_node_class
+class SqPNorm(RBFCost):
+  """Squared p-norm of the difference of two vectors.
+
+  For details on the derivation of the Legendre transform of the norm, see e.g.
+  the reference :cite:`boyd:04`, p.93/94.
+  https://web.stanford.edu/~boyd/cvxbook/bv_cvxbook.pdf
+  """
+
+  def __init__(self, p: float):
+    assert p > 1.0
```

Comment:
Please add some message.

Reply:
Actually I reconsidered the >. 

I added the option to have an =.

What will happen is that the Brenier theorem will be completely degenerate (involving the inf. norm in both cases, whose gradient is also degenerate). One of the nice things in JAX is that this will still work (the grad will be 0 in all coordinates except for the max, where it will be 1), even if it makes little sense.

I also added the p-norm to power p.